### DISCUSSION TITLE
Add keybinding for insert mode

### DISCUSSION THREAD

**Discussion by bgschiller** [LazyVim/LazyVim#4088] [Q&A] [answered]
I'd like to map `<C-j>`, `<C-k>`, etc to a lua command even when I'm in insert mode. For example, I'd like `<C-h>` to run `require("smart-splits").move_cursor_left()`. I have this working in normal mode with the following:

```lua
return {
  "mrjones2014/smart-splits.nvim",
  build = "./kitty/install-kittens.bash",
  keys = {
    {
      "<C-h>",
      function()
        require("smart-splits").move_cursor_left()
      end,
      desc = "Move to left window",
    },
    // ... similar for other directions
  }
}
```

But I can't figure out how to make it work in insert mode. I tried

```lua
vim.keymap.set("i", "<C-j>", "<Esc> <C-j>", { desc = "Go to below window", remap = true })
// same for other directions
```

but that doesn't seem to do anything.

**Comment by dotfrag** [comment, marked as answer]
Do you have `C-j` in normal mode mapped to something?

**Reply by bgschiller** [reply]
I do, yes. It's part of the text I omitted in the first code snippet. `<C-j>` is working for me in normal mode.

**Reply by bgschiller** [reply]
Wait, I just retried and it's working now 🤦 

Something else must have been broken when I was last testing it.

**Reply by dotfrag** [reply]
I added this to my config and it's not working, is it supposed to move around in kitty splits?

```lua
    {
      "mrjones2014/smart-splits.nvim",
      build = "./kitty/install-kittens.bash",
      -- stylua: ignore
      keys = {
        { "<C-h>", function() require("smart-splits").move_cursor_left() end, desc = "Move to left window" },
        { "<C-j>", function() require("smart-splits").move_cursor_down() end, desc = "Move to left window" },
        { "<C-k>", function() require("smart-splits").move_cursor_up() end, desc = "Move to left window" },
        { "<C-l>", function() require("smart-splits").move_cursor_right() end, desc = "Move to left window" },
      },
    },
  },
```

**Reply by dotfrag** [reply]
I believe you do not need `vim.keymap.set`, instead try adding the modes in `keys` like it's done here 

https://github.com/LazyVim/LazyVim/blob/c1b76ee235a2cccff6370ecfca57bdacd5fe6258/lua/lazyvim/plugins/coding.lua#L105-L124

and here

https://github.com/LazyVim/LazyVim/blob/c1b76ee235a2cccff6370ecfca57bdacd5fe6258/lua/lazyvim/plugins/editor.lua#L151-L157

**Reply by bgschiller** [reply]
There's some setup required in kitty.conf as well. It lets me use the same motion commands to move between vim windows and kitty windows. Here's a video of what I mean https://www.loom.com/share/049108333df445d194b8d13c5c121d88

**Reply by bgschiller** [reply]
Ah, I didn't know about the `mode` keyword arg. I'll try that instead, thanks!

**Reply by dotfrag** [reply]
Interesting, thanks for the video. I don't use kitty splits to go into the effort of troubleshooting, but I hope the `mode` key will work for you, good luck :)

**Reply by bgschiller** [reply]
Thanks! I'm actually still having the issue that appears in the video: when I first start up neovim, the `<C-l>`, `<C-h>`, etc keys move me between kitty windows but not vim windows. If I select the same command via the keymap list `<space>sk`, it will work that time and every subsequent time.

Any ideas? I can open a new question, since it's a bit of a different question.

**Reply by dotfrag** [reply]
So the first time you press `<C-l>` or `<C-h>` it changes kitty focus, even if there's a vim window available to go to. What happens if you go back to neovim pane and try the keymap again? (don't use `<leader>sk`)

**Reply by bgschiller** [reply]
I can go back and forth any number of times and it still takes me between ktity windows. Here's a video again with more details https://www.loom.com/share/28666eef5b7d42bd840e5c0da66b3674

**Reply by dotfrag** [reply]
Loom asks me to sign in and I can't seem to be able to bypass it :S

I asked you to test this because I was wondering if something is off with the plugin loading and the setting of IS_NVIM variable. But since you are experiencing this, it doesn't seem to be the case.

Just to make sure, can you add this to `autocmds.lua`

```lua
vim.api.nvim_create_autocmd({ "VimEnter", "VimResume" }, {
    group = vim.api.nvim_create_augroup("KittySetVarVimEnter", { clear = true }),
    callback = function()
        io.stdout:write("\x1b]1337;SetUserVar=IS_NVIM=MQo\007")
    end,
})

vim.api.nvim_create_autocmd({ "VimLeave", "VimSuspend" }, {
    group = vim.api.nvim_create_augroup("KittyUnsetVarVimLeave", { clear = true }),
    callback = function()
        io.stdout:write("\x1b]1337;SetUserVar=IS_NVIM\007")
    end,
})
```

EDIT: Fixed variable name.

EDIT 2: If you do not use `<leader>sk`, and instead manually type the command `:lua require('smart-splits').move_cursor_left`, what happens?

Try with and without `()`:

```
:lua require('smart-splits').move_cursor_left
```

```
:lua require('smart-splits').move_cursor_left()
```

**Reply by bgschiller** [reply]
Ah, sorry about that; I fixed the loom link to allow viewing by anyone

**Reply by bgschiller** [reply]
>  If you do not use <leader>sk, and instead manually type the command :lua require('smart-splits').move_cursor_left, what happens?

It also begins to work in that case.

The `IS_NVIM` setting seems to be working, and I think we'd see different behavior if kitty was eating those keypresses.

**Reply by bgschiller** [reply]
> Try with and without ():

I get an error without the ()
```
   Error  10:50:33 msg_show.lua_error   require("smart-splits").move_cursor_left E5107: Error loading lua [string ":lua"]:1: '=' expected near '<eof>'
```

**Reply by dotfrag** [reply]
> I get an error without the ()

Okay no problem, I was a bit confused from the plugin's readme.

> The IS_NVIM setting seems to be working, and I think we'd see different behavior if kitty was eating those keypresses.

Does that "solve" the problem? If that's the case then it's related to plugin loading.

**Reply by bgschiller** [reply]
I'm afraid it doesn't solve it.

**Reply by dotfrag** [reply]
I'm out of ideas.. by looking at the video, it looks like the plugin doesn't set the `IS_NVIM` variable until you run the command.. Do you have anything else in the plugin spec besides `build` and `keys` that might affect how the plugin is loaded?

**Reply by bgschiller** [reply]
No, I'll paste the whole thing below just in case you see something.

I'm doing a bit of research on the kitty user-var thing, to see how to inspect those values. I had assumed they were environment variables, but I guess they're something different. What do the escape codes do in the autocmd snippet you gave me?

**Reply by bgschiller** [reply]
```lua
return {
  "mrjones2014/smart-splits.nvim",
  build = "./kitty/install-kittens.bash",
  keys = {
    {
      "<C-h>",
      function()
        require("smart-splits").move_cursor_left()
      end,
      mode = { "i", "n", "v" },
      desc = "Move to left window",
    },
    {
      "<C-j>",
      function()
        require("smart-splits").move_cursor_down()
      end,
      mode = { "i", "n", "v" },
      desc = "Move to below window",
    },
    {
      "<C-k>",
      function()
        require("smart-splits").move_cursor_up()
      end,
      desc = "Move to above window",
    },
    {
      "<C-l>",
      function()
        require("smart-splits").move_cursor_right()
      end,
      mode = { "i", "n", "v" },
      desc = "Move to right window",
    },
    {
      "<C-\\>",
      function()
        require("smart-splits").move_cursor_previous()
      end,
      mode = { "i", "n", "v" },
      desc = "Move cursor to previous window",
    },
  },
}
```

**Reply by dotfrag** [reply]
Thanks, it looks normal.. see my response below (new comment).

**Comment by folke** [comment]
LazyVim's terminal by default maps `ctrl+hjkl` to the same keys to switch windows.

You can disable that by remapping to:

```lua
  LazyVim.terminal(nil, {
    -- esc_esc = false,
    ctrl_hjkl = false,
  })
```

Edit: this has nothing to do with your questions

**Comment by dotfrag** [comment]
@bgschiller  Please continue the conversation here, I'm lost with the other one.

LazyVim has defaults for `c-h` etc. here:

https://github.com/LazyVim/LazyVim/blob/c1b76ee235a2cccff6370ecfca57bdacd5fe6258/lua/lazyvim/config/keymaps.lua#L13-L17

Did you try disabling them in case they conflict (they might prevent the plugin from loading/setting the variable)?

In your `keymaps.lua`:

```lua
vim.keymap.del("n", "<C-h>")
vim.keymap.del("n", "<C-j>")
vim.keymap.del("n", "<C-k>")
vim.keymap.del("n", "<C-l>")
```

Can you try adding this in the plugin spec:

```lua
return {
  "mrjones2014/smart-splits.nvim",
   ....
   config = function(_, opts)
      require('smart-splits').setup(opts)
   end
}
```

Can you also please confirm whether the autocommands I posted above work or not? I didn't understand from the previous responses if you tried it.

**Reply by dotfrag** [reply]
Also can you explicitly disable lazy loading just to test:

```lua
lazy = false,
```

in the plugin spec.

**Reply by bgschiller** [reply]
Yes, I tried the autocommands and they didn't seem to work.

I've tried deleting the keymaps and adding the config option to the plugin spec and those didn't work either.

One idea I had: Could I run `require("smart-splits")` when nvim first starts up? That seems to be enough to kick it into working—I don't need to also call a `.move_cursor_left()` function. I tried adding this to `lua/config/options.lua` but I get an error saying that the module wasn't found.

**Reply by bgschiller** [reply]
Ah, just saw your message. I'll try that, I wasn't sure how to do it :)

**Reply by bgschiller** [reply]
`lazy = false` did the trick! 

Thanks for spending so much time working through this with me. I'm planning to write a blog post on this, since there were a couple of other gotchas I ran into while setting it up. Is there a website or URL I can link to for you to give you some credit?

**Reply by dotfrag** [reply]
Dude, I just realised something, in the plugin spec, you have `keys`, but neither `opts`, nor `config`. This causes the plugin to not load.

Can you remove `lazy=false` and just add `opts = {}`?

**Reply by bgschiller** [reply]
Hmm, replacing `lazy = false` with `opts = {}` brings it back to not working. :shrug:

**Reply by dotfrag** [reply]
That's strange, setting `opts` with no event should make the plugin load.. I think?

Imo the ideal setup would be:

```lua
event = "LazyFile", -- or "VeryLazy
opts = {},
keys = ...
```

Anyhow, if `lazy =false` worked then I guess we can leave it at that.

> Is there a website or URL I can link to for you to give you some credit?

No but thank you, I'm glad to help!
